### ISSUE TITLE
How to make training sessions reproducible?

### ISSUE THREAD

**Issue by manacoa** [DLR-RM/stable-baselines3#292]
### Question

Are there any guidelines for how to make training sessions deterministic and reproducible? I'm training OpenAI Gym MuJoCo environments with PPO and my training graphs are different each time. Not sure if this is something I'm doing wrong or if it's simply impossible. A checklist would be very useful.

### Additional context

What I've tried before each training session:

    seed = 0
    env.seed(seed)
    env.action_space.seed(seed)
    stable_baselines3.common.utils.set_random_seed(seed, using_cuda=True)
 
### Checklist

- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Hey. See [docs](https://stable-baselines3.readthedocs.io/en/master/guide/algos.html#reproducibility) on this. You also need to set the seed of the model with `model = PPO(..., seed=0)`. This _should_ do the trick, but if it does not, please share full code the replicate the issue with a toy-task (e.g. CartPole).

**Comment by manacoa** [comment]
That did the trick! Thanks, and apologies for missing it.

**Comment by axkoenig** [comment]
Hi, 

I have two follow up question on this. I have a custom environment and set the seeds as described above (including setting the seed for my TD3 model). By looking at the plots from my different training runs, they are similar, but not the same. But I don't care too much about this as long as it solves the task. 
![Screen Shot 2021-03-23 at 16 11 36](https://user-images.githubusercontent.com/50868646/112211573-71a64280-8bf2-11eb-8f80-cdf73770e800.png)

1. My agent learns to solve the task very well by the end of the training procedure. At the end of training I save the model. When I load the model and test it using exactly the same task, my agent behaves differently (and worse) compared to at the end of the training procedure. What could be the cause of this?
2. Are there any downsides of setting a seed? 

Cheers, 
Alex

**Comment by Miffyli** [comment]
@axkoenig 

If you are using a GPU, PyTorch might use non-deterministic algorithms for this ([see PyTorch's docs on this](https://pytorch.org/docs/stable/notes/randomness.html)). Same may apply to running on CPU. 

1) See #326 for related a issue (but this contains training). Seed should not affect this. Also make sure you use same normalization statistics for evaluation environment and use right `deterministic` parameter for the `predict` function (`False` if you want to match training). @araffin We should probably include these tricks in docs somewhere.

2) Not really, it just makes sure that all random numbers should be same between runs. Only downside I see is that if you fix your seed and keep tuning your hyperparameters, you might overfit to that one specific seed (hence, for proper evaluation, you should run training on multiple different seeds).

**Comment by araffin** [comment]
>By looking at the plots from my different training runs, they are similar, but not the same

I would say something is wrong in seeding the process then (probably from the env), even using a GPU (I did some quick tests on pendulum using the zoo and could get perfectly reproducible results).
We have unit tests to check that (on cpu):
https://github.com/DLR-RM/stable-baselines3/blob/master/tests/test_deterministic.py

>When I load the model and test it using exactly the same task and the same seed, my agent behaves differently (and worse) compared to at the end of the training procedure.

As @Miffyli wrote, could be a problem of normalization statistics.

That's why I would recommend using the rl zoo (https://github.com/DLR-RM/rl-baselines3-zoo) which ensure that everything is the same between training and testing.

Another things: during training, the TD3 agent is using an exploration policy (deterministic controller + noise) and the exploration noise is removed at test time. You can manually add it at test time if you want it, but for continuous control it should not be the case (if you need it, it may mean that you probably need more training, so that the controller does not need noise to converge to the optimal behavior).

>Are there any downsides of setting a seed?

To add on @Miffyli remark, when you set the seed, pytorch will try to use deterministic computation, which can be slightly slower than the default. But to be honest, I did not experience any significant slowdown yet.
